### PR TITLE
Make OptimizerConfig a trait (#4631) (#4638)

### LINKED ISSUE #4631: Make OptimizerConfig a Trait
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Similar to #4629 

**Describe the solution you'd like**

The proliferation of various different config structs (#4349) is confusing, not to mention error prone (#4630)

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #4638: Filter Null Keys Update Not Taking Effect
**Describe the bug**

The filter_null_keys option is applied during the construction of `Optimizer`, at the same time as the `SessionContext` is constructed. 

Consequently if the value is updated, it will not take effect

**To Reproduce**

**Expected behavior**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4631 
Closes #4638 

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/execution/context.rs
Comment (tustvold) [inline datafusion/core/src/execution/context.rs:1560]: Moving this into the `OptimizerConfig` passed at optimize time is the fix for #4638

File: datafusion/optimizer/src/optimizer.rs
Comment (tustvold) [inline datafusion/optimizer/src/optimizer.rs:183]: This is the other half of the fix for #4638 - determine the rules dynamically

File: datafusion/optimizer/src/optimizer.rs
Comment (tustvold) [inline datafusion/optimizer/src/optimizer.rs:82]: I felt this formulation was more future proof than adding a specific function for enabling/disabling `FilterNullJoins`

File: datafusion/optimizer/src/optimizer.rs
Comment (tustvold) [inline datafusion/optimizer/src/optimizer.rs:118]: We use 1, as previously the implementation returned the value post-increment which the atomics don't provide

File: datafusion/optimizer/src/optimizer.rs
Comment (tustvold) [inline datafusion/optimizer/src/optimizer.rs:104]: I think this limited interior mutability is better than the potential confusion around rules taking a mutable `OptimizerConfig`

File: datafusion/core/src/execution/context.rs
Comment (tustvold) [inline datafusion/core/src/execution/context.rs:1736]: This is the ultimate motivation for this change, to allow a single config container, that then just implements the traits needed by the various sub-systems

**Comment by jackwener** [comment]
I have some PR about `optimizer`😂, look like it will make some conflict.

File: datafusion/core/src/execution/context.rs
Comment (alamb) [reply datafusion/core/src/execution/context.rs:1736]: I think having the sub crates depend on a traits that are implemented in the core trait sounds like a good idea to me

File: datafusion/optimizer/src/optimizer.rs
Comment (alamb) [reply datafusion/optimizer/src/optimizer.rs:94]: It is also good that `OptimizerContext` doesn't implement `Clone` which will discourage making copies that could get out of date

**Review by alamb — APPROVED** [review]
Looks like a (really) nice improvement to me -- thank you @tustvold 🏆 

I think we should gather some more input and leave this open a while before merging it even though I think the actual required API change downstream are relatively low

cc @andygrove @jackwener @yahoNanJing @Dandandan @thinkharderdev @xudong963

**Review by thinkharderdev — APPROVED** [review]
Makes sense to me

File: datafusion/optimizer/src/optimizer.rs
Comment (waynexia) [reply datafusion/optimizer/src/optimizer.rs:94]: Looks good 👍

**Comment by tustvold** [comment]
I plan to merge this in the next few hours unless anybody objects / needs more time to review

**Comment by alamb** [comment]
🎉 

really nice work @tustvold

**Pull request merged**
